lib/vector-core/src/config/mod.rs

```diff
@@ -100,42 +100,129 @@ impl Input {
-    /// output consumers will receive if they declare the component itself as an input.
-    pub fn default(ty: DataType) -> Self {
+    /// Asserts that `ty` contains [`DataType::Log`].
+    #[must_use]
+    pub fn new_logs(ty: DataType, schema_definition: schema::Definition) -> Self {
+        assert!(ty.contains(DataType::Log));
+
         Self {
             port: None,
             ty,
-            log_schema_definition: None,
+            schema_definition: Some(schema_definition),
+        }
+    }
+
+    /// Create a `SourceOutput` of the given data type that contains no output `Definition`s.
+    /// Designed for use in metrics sources.
+    ///
+    /// Sets the datatype to be [`DataType::Metric`].
+    #[must_use]
+    pub fn new_metrics() -> Self {
+        Self {
+            port: None,
+            ty: DataType::Metric,
+            schema_definition: None,
+        }
+    }
+
+    /// Create a `SourceOutput` of the given data type that contains no output `Definition`s.
+    /// Designed for use in trace sources.
+    ///
+    ///
+    /// Sets the datatype to be [`DataType::Trace`].
+    #[must_use]
+    pub fn new_traces() -> Self {
+        Self {
+            port: None,
+            ty: DataType::Trace,
+            schema_definition: None,
         }
     }
 
-    /// Set the schema definition for this `Output`.
+    /// Return the schema [`schema::Definition`] from this output.
+    ///
+    /// Takes a `schema_enabled` flag to determine if the full definition including the fields
+    /// and associated types should be returned, or if a simple definition should be returned.
+    /// Schema enabled is set in the users configuration.
     #[must_use]
-    pub fn with_schema_definition(mut self, schema_definition: schema::Definition) -> Self {
-        self.log_schema_definition = Some(schema_definition);
+    pub fn schema_definition(&self, schema_enabled: bool) -> Option<schema::Definition> {
+        self.schema_definition.as_ref().map(|definition| {
+            if schema_enabled || definition.log_namespaces().contains(&LogNamespace::Vector) {
+                definition.clone()
+            } else {
+                let mut new_definition =
+                    schema::Definition::default_for_namespace(definition.log_namespaces());
+                new_definition.add_meanings(definition.meanings());
+                new_definition
+            }
+        })
+    }
+
+    /// Consumes the object and returns the schema [`schema::Definition`] from this output.
+    ///
+    /// Takes a `schema_enabled` flag to determine if the full definition including the fields
+    /// and associated types should be returned, or if a simple definition should be returned.
+    /// Schema enabled is set in the user's configuration.
+    #[must_use]
+    pub fn into_schema_definition(self, schema_enabled: bool) -> Option<schema::Definition> {
+        self.schema_definition.map(|definition| {
```

Comment:
This seems like it can re-use the `schema_definition()` implementation?

Reply:
The combination of the `as_ref` and `clone` in one and not the other makes it annoyingly different enough that I can't really see a way around it. 

Granted, none of this is in performance sensitive code so I could remove `into_schema_definition` completely and be ok with the extra clone if I had to.

Reply:
Yea, I would personally prefer to add extra clones to simplify the code since this is not performance sensitive.